common/persistence/client/fault_injection.go

```diff
@@ -0,0 +1,925 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/persistence"
+)
+
+type (
+	FaultInjectionDataStoreFactory struct {
+		baseFactory    DataStoreFactory
+		config         *config.FaultInjection
+		errorGenerator ErrorGenerator
+	}
+
+	// FaultInjectionShardStore is a lower level of ShardManager
+	FaultInjectionShardStore struct {
+		baseShardStore persistence.ShardStore
+		config         *config.FaultInjection
+		errorGenerator ErrorGenerator
+	}
+
+	// FaultInjectionTaskStore is a lower level of TaskManager
+	FaultInjectionTaskStore struct {
+		baseTaskStore  persistence.TaskStore
+		config         *config.FaultInjection
+		errorGenerator ErrorGenerator
+	}
+
+	// FaultInjectionMetadataStore is a lower level of MetadataManager
+	FaultInjectionMetadataStore struct {
+		baseMetadataStore persistence.MetadataStore
+		config            *config.FaultInjection
+		errorGenerator    ErrorGenerator
+	}
+
+	// FaultInjectionClusterMetadataStore is a lower level of ClusterMetadataManager.
+	// There is no Internal constructs needed to abstract away at the interface level currently,
+	//  so we can reimplement the ClusterMetadataManager and leave this as a placeholder.
+	FaultInjectionClusterMetadataStore struct {
+		baseCMStore    persistence.ClusterMetadataStore
+		config         *config.FaultInjection
+		errorGenerator ErrorGenerator
+	}
+
+	// FaultInjectionExecutionStore is used to manage workflow execution including mutable states / history / tasks.
+	FaultInjectionExecutionStore struct {
+		baseExecutionStore persistence.ExecutionStore
+		config             *config.FaultInjection
+		errorGenerator     ErrorGenerator
+	}
+
+	// FaultInjectionQueue is a store to enqueue and get messages
+	FaultInjectionQueue struct {
+		baseQueue      persistence.Queue
+		config         *config.FaultInjection
+		errorGenerator ErrorGenerator
+	}
+
+	ErrorGenerator interface {
+		Generate() error
+	}
+
+	DefaultErrorGenerator struct {
+		rate float64
+		r    *rand.Rand
+	}
+
+	NoopErrorGenerator struct{}
+)
+
+func NewDefaultErrorGenerator(rate float64) *DefaultErrorGenerator {
+	return &DefaultErrorGenerator{
+		rate: rate,
+		r:    rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (p *DefaultErrorGenerator) Generate() error {
+	if p.rate <= 0 {
+		return nil
+	}
+
+	if roll := p.r.Float64(); roll < p.rate {
+		msg := fmt.Sprintf("FaultInjection injected error with rate %f, roll: %f", p.rate, roll)
+		errTypeRoll := p.r.Float64()
+		if errTypeRoll < 0.1 {
+			return serviceerror.NewInternal(msg)
+		}
+		return &persistence.TimeoutError{Msg: msg}
```

Comment:
can this be configurable? 
e.g. initialization logic will provide a list of errors to return, with corresponding distribution

Reply:
@Ardagan I think we agreed to make error types (and their likely-hood ) configurable. It it planed in your follow up PRs?

Reply:
At least, we want to mimic a few other error types, like OwnershipLostError etc.

Reply:
Seems that OwnershipLostError  mostly sums it up. Updating.